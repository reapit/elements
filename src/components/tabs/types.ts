import { HTMLAttributes, InputHTMLAttributes, LabelHTMLAttributes } from 'react'

export interface TabsBaseProps extends HTMLAttributes<HTMLDivElement | HTMLLabelElement | HTMLSpanElement> {}

export interface TabsInputProps extends InputHTMLAttributes<HTMLInputElement> {}

export interface TabsLabelProps extends LabelHTMLAttributes<HTMLLabelElement> {}

export interface TabsOption {
  id: string
  value: string
  text: string
  isChecked: boolean
}

export interface TabsProps extends HTMLAttributes<HTMLInputElement> {
  options: TabsOption[]
  name: string
  isControlled?: boolean
  isFullWidth?: boolean
  hasNoBorder?: boolean
}
