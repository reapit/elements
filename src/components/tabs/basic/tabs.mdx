import { Story, Canvas, Controls, Meta, Source } from '@storybook/blocks'
import { Tabs } from '../index'
import { RenderHtmlMarkup } from '../../../storybook/render-html-markup'
import * as TabsStories from './tabs.stories'

<Meta of={TabsStories} />

## Multi Tabs - Tabs

The `Tabs` component uses an underlying radio button to manage toggling without JavaScript. As such, you can manage state using any Form library or by referencing the `onChange` input event.

As always, for non React users, you can just apply the classes in the rendered markup to your input and label elements.

Typically you will use this component as a tertiary in-page navigation.

<Canvas of={TabsStories.BasicUsage} />

<RenderHtmlMarkup of={TabsStories.BasicUsage} />

## Tabs - Full Width

Slight variation on the above where the Tabs spans the full width of it's container.

<Canvas of={TabsStories.FullWidth} />

<RenderHtmlMarkup of={TabsStories.FullWidth} />

## Tabs - No border

Another variation on the above where the Tabs have no bottom border.

<Canvas of={TabsStories.NoBorder} />

<RenderHtmlMarkup of={TabsStories.NoBorder} />
