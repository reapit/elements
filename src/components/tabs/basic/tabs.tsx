import { FC } from 'react'
import { cx } from '@linaria/core'
import { ElTab, ElTabsLabel, ElTabsWrap, ElTabsFooter, ElTabsOptionsWrap, ElTabsItem } from '../__styles__'
import { TabsBaseProps, TabsInputProps, TabsLabelProps } from '../types'

export const Tab: FC<TabsInputProps> = ({ children, className, ...rest }) => (
  <ElTab className={cx(className)} {...rest}>
    {children}
  </ElTab>
)

export const TabsWrap: FC<TabsBaseProps> = ({ children, className, ...rest }) => (
  <ElTabsWrap className={cx(className)} {...rest}>
    {children}
  </ElTabsWrap>
)

export const TabsLabel: FC<TabsLabelProps> = ({ children, className, ...rest }) => (
  <ElTabsLabel className={cx(className)} {...rest}>
    {children}
  </ElTabsLabel>
)

export const TabsFooter: FC<TabsBaseProps> = ({ children, className, ...rest }) => (
  <ElTabsFooter className={cx(className)} {...rest}>
    {children}
  </ElTabsFooter>
)

export const TabsOptionsWrap: FC<TabsBaseProps> = ({ children, className, ...rest }) => (
  <ElTabsOptionsWrap className={cx(className)} {...rest}>
    {children}
  </ElTabsOptionsWrap>
)

export const TabsItem: FC<TabsBaseProps> = ({ children, className, ...rest }) => (
  <ElTabsItem className={cx(className)} {...rest}>
    {children}
  </ElTabsItem>
)
