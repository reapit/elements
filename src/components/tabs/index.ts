export * from './__styles__'
export * from './types'
export * from './tabs.atoms'
export * from './tabs'
