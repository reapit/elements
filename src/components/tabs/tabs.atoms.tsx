import { FC, ForwardedRef, forwardRef, InputHTMLAttributes, LegacyRef } from 'react'
import { cx } from '@linaria/core'
import { ElTab, ElTabsLabel, ElTabsWrap, ElTabsFooter, ElTabsOptionsWrap, ElTabsItem } from './styles'
import { TabsBaseProps, TabsInputProps, TabsLabelProps } from './types'

export const Tab: TabsInputProps = forwardRef(
  ({ children, className, ...rest }, ref: ForwardedRef<InputHTMLAttributes<HTMLInputElement>>) => (
    <ElTab className={cx(className)} {...rest} ref={ref as unknown as LegacyRef<HTMLInputElement>}>
      {children}
    </ElTab>
  ),
)

export const TabsWrap: FC<TabsBaseProps> = ({ children, className, ...rest }) => (
  <ElTabsWrap className={cx(className)} {...rest}>
    {children}
  </ElTabsWrap>
)

export const TabsLabel: FC<TabsLabelProps> = ({ children, className, ...rest }) => (
  <ElTabsLabel className={cx(className)} {...rest}>
    {children}
  </ElTabsLabel>
)

export const TabsFooter: FC<TabsBaseProps> = ({ children, className, ...rest }) => (
  <ElTabsFooter className={cx(className)} {...rest}>
    {children}
  </ElTabsFooter>
)

export const TabsOptionsWrap: FC<TabsBaseProps> = ({ children, className, ...rest }) => (
  <ElTabsOptionsWrap className={cx(className)} {...rest}>
    {children}
  </ElTabsOptionsWrap>
)

export const TabsItem: FC<TabsBaseProps> = ({ children, className, ...rest }) => (
  <ElTabsItem className={cx(className)} {...rest}>
    {children}
  </ElTabsItem>
)
