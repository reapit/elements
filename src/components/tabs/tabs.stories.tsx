import { Tabs } from './index'
import { useState } from 'react'
import { InputGroup } from '../input-group'
import { TextBase } from '../typography'

export default {
  title: 'Tabs',
  component: Tabs,
}

export const TabsDefault = {
  render: ({}) => (
    <Tabs
      name="my-cool-tabs-default"
      options={[
        {
          id: 'tab-1',
          value: 'tab-1',
          text: 'Tab Content 1',
          isChecked: true,
        },
        {
          id: 'tab-2',
          value: 'tab-2',
          text: 'Tab Content 2',
          isChecked: false,
        },
        {
          id: 'tab-3',
          value: 'tab-3',
          text: 'Tab Content 3',
          isChecked: false,
        },
      ]}
    />
  ),
  name: 'Tabs - Default',
}

export const TabsFullWidth = {
  render: ({}) => (
    <Tabs
      name="my-cool-tabs-full-width"
      isFullWidth
      options={[
        {
          id: 'tab-1-fw',
          value: 'tab-1-fw',
          text: 'Tab Content 1',
          isChecked: true,
        },
        {
          id: 'tab-2-fw',
          value: 'tab-2-fw',
          text: 'Tab Content 2',
          isChecked: false,
        },
        {
          id: 'tab-3-fw',
          value: 'tab-3-fw',
          text: 'Tab Content 3',
          isChecked: false,
        },
      ]}
    />
  ),
  name: 'Tabs - Full Width',
}

export const TabsReactExampleUncontrolled = {
  render: ({}) => {
    const [tab, setTab] = useState('1')

    return (
      <>
        <Tabs
          name="my-cool-tabs-react"
          isFullWidth
          onChange={(e: any) => setTab(e.target.value)}
          options={[
            {
              id: 'tab-1-react',
              value: '1',
              text: 'Tab Content 1',
              isChecked: tab === '1',
            },
            {
              id: 'tab-2-react',
              value: '2',
              text: 'Tab Content 2',
              isChecked: tab === '2',
            },
            {
              id: 'tab-3-react',
              value: '3',
              text: 'Tab Content 3',
              isChecked: tab === '3',
            },
          ]}
        />
        {tab === '1' && <TextBase hasGreyText>Tab 1 Content</TextBase>}
        {tab === '2' && <TextBase hasGreyText>Tab 2 Content</TextBase>}
        {tab === '3' && <TextBase hasGreyText>Tab 3 Content</TextBase>}
        <InputGroup
          value={tab}
          label="The value of this input will change with the state, but updating the state of the input will not change the tab."
          type="number"
          min="1"
          max="3"
          onChange={(e) => {
            setTab(e.target.value)
          }}
        />
      </>
    )
  },
  name: 'Tabs - React Example uncontrolled',
}

export const TabsReactExampleControlled = {
  render: ({}) => {
    const [tab, setTab] = useState('1')

    return (
      <>
        <Tabs
          name="my-cool-tabs-react-controlled"
          isFullWidth
          isControlled
          onChange={(e: any) => setTab(e.target.value)}
          options={[
            {
              id: 'tab-1-react-controlled',
              value: '1',
              text: 'Tab Content 1',
              isChecked: tab === '1',
            },
            {
              id: 'tab-2-react-controlled',
              value: '2',
              text: 'Tab Content 2',
              isChecked: tab === '2',
            },
            {
              id: 'tab-3-react-controlled',
              value: '3',
              text: 'Tab Content 3',
              isChecked: tab === '3',
            },
          ]}
        />
        {tab === '1' && <TextBase hasGreyText>Tab 1 Content</TextBase>}
        {tab === '2' && <TextBase hasGreyText>Tab 2 Content</TextBase>}
        {tab === '3' && <TextBase hasGreyText>Tab 3 Content</TextBase>}
        <InputGroup
          value={tab}
          label="The value of this input will change with the state and updating the input will change the tab"
          type="number"
          min="1"
          max="3"
          onChange={(e) => {
            setTab(e.target.value)
          }}
        />
      </>
    )
  },
  name: 'Tabs - React Example controlled',
}
