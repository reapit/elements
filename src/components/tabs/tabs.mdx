import { Story, Canvas, Controls, Meta, Source } from '@storybook/blocks'
import { Tabs } from './index'
import { RenderHtmlMarkup } from '../../storybook/render-html-markup'
import * as TabsStories from './tabs.stories'

<Meta of={TabsStories} />

## Multi Tabs - Tabs

The `Tabs` component uses an underlying radio button to manage toggling without JavaScript. As such, you can manage state using any Form library or by referencing the `onChange` input event.

For react users, the component does not render children, rather it accepts a list of options and maps through them on render.

As always, for non React users, you can just apply the classes in the rendered markup to your input and label elements.

Typically you will use this component as a tertiary in-page navigation.

Props available for the React version of the component;

<Controls />

Where `TabsOption` is defined as;

<Source code={`
export interface TabsOption {
  id: string
  value: string
  text: string
  isChecked: boolean
}
`} language="tsx" />

<Canvas of={TabsStories.TabsDefault} />

<RenderHtmlMarkup of={TabsStories.TabsDefault} />

## Tabs - Full Width

Slight variation on the above where the Tabs spans the full width of it's container.

<Canvas of={TabsStories.TabsFullWidth} />

<RenderHtmlMarkup of={TabsStories.TabsFullWidth} />


## Tabs - React Example uncontrolled

Simple example demonstrating how to transition content using tabs. In practice, you may want to dispatch to a router or similar.

The underlying radio buttons are uncontrolled by default so the `isChecked` is used to set the `defaultChecked` in this example. This is to allow the radio to behave nativele i.e. it will update the state using the `onChange` event but not vice versa.

<Canvas of={TabsStories.TabsReactExampleUncontrolled} />
<RenderHtmlMarkup of={TabsStories.TabsReactExampleUncontrolled} />

## Tabs - React Example controlled

Simple example demonstrating how to transition content using tabs. In practice, you may want to dispatch to a router or similar.

The underlying radio buttons are uncontrolled by default so the `isChecked` is used to set the `defaultChecked` in this example. This is to allow the radio to behave nativele i.e. it will update the state using the `onChange` event but not vice versa.

<Canvas of={TabsStories.TabsReactExampleControlled} />
<RenderHtmlMarkup of={TabsStories.TabsReactExampleControlled} />
