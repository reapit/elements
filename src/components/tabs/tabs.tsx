import { FC, Fragment, MutableRefObject, useRef } from 'react'
import { cx } from '@linaria/core'
import { elTabsFullWidth, elTabsHasNoBorder } from './styles'
import { handleKeyboardEvent } from '../../storybook/handle-keyboard-event'
import { TabsProps } from './types'
import { TabsWrap, TabsOptionsWrap, Tab, TabsLabel, TabsItem, TabsFooter } from './tabs.atoms'

export const handleKeyboardTabChange =
  (tabsRefs: MutableRefObject<(HTMLInputElement | null)[]>, index: number) => () => {
    tabsRefs.current[index]?.click()
  }

export const Tabs: FC<TabsProps> = ({ className, isFullWidth, hasNoBorder, isControlled, name, options, ...rest }) => {
  const tabsRefs = useRef<(HTMLInputElement | null)[]>([])

  return (
    <TabsWrap className={cx(className, isFullWidth && elTabsFullWidth)}>
      <TabsOptionsWrap role="tablist">
        {options.map(({ id, value, text, isChecked }, index) => (
          <Fragment key={id}>
            <Tab
              id={id}
              ref={(el) => ((tabsRefs as any).current[index] = el)}
              name={name}
              value={value}
              type="radio"
              aria-selected={isChecked}
              {...rest}
              checked={isControlled ? isChecked : undefined}
              defaultChecked={isControlled ? undefined : isChecked}
            />
            <TabsLabel
              htmlFor={id}
              role="tab"
              tabIndex={0}
              onKeyDown={handleKeyboardEvent('Enter', handleKeyboardTabChange(tabsRefs, index))}
            >
              <TabsItem>{text}</TabsItem>
            </TabsLabel>
          </Fragment>
        ))}
      </TabsOptionsWrap>
      <TabsFooter className={cx(isFullWidth && elTabsFullWidth, hasNoBorder && elTabsHasNoBorder)} />
    </TabsWrap>
  )
}
