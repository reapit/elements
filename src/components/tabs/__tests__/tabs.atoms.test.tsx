import { render } from '@testing-library/react'
import { Tab, TabsFooter, TabsItem, TabsLabel, TabsOptionsWrap, TabsWrap } from '../tabs.atoms'

describe('Tab', () => {
  it('should match a snapshot and render children', () => {
    const wrapper = render(<Tab className="foo" />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('Tab', () => {
  it('should match a snapshot and render children', () => {
    const wrapper = render(<Tab className="foo" />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('TabsWrap', () => {
  it('should match a snapshot and render children', () => {
    const wrapper = render(
      <TabsWrap className="foo">
        <div>Child</div>
      </TabsWrap>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('TabsLabel', () => {
  it('should match a snapshot and render children', () => {
    const wrapper = render(
      <TabsLabel className="foo">
        <div>Child</div>
      </TabsLabel>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('TabsFooter', () => {
  it('should match a snapshot and render children', () => {
    const wrapper = render(
      <TabsFooter className="foo">
        <div>Child</div>
      </TabsFooter>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('TabsOptionsWrap', () => {
  it('should match a snapshot and render children', () => {
    const wrapper = render(
      <TabsOptionsWrap className="foo">
        <div>Child</div>
      </TabsOptionsWrap>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('TabsItem', () => {
  it('should match a snapshot and render children', () => {
    const wrapper = render(
      <TabsItem className="foo">
        <div>Child</div>
      </TabsItem>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})
