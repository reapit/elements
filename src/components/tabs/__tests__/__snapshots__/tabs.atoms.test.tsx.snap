// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tab > should match a snapshot and render children 1`] = `
<DocumentFragment>
  <input
    class="foo el-tab"
  />
</DocumentFragment>
`;

exports[`Tab > should match a snapshot and render children 2`] = `
<DocumentFragment>
  <input
    class="foo el-tab"
  />
</DocumentFragment>
`;

exports[`TabsFooter > should match a snapshot and render children 1`] = `
<DocumentFragment>
  <div
    class="foo el-tabs-footer"
  >
    <div>
      Child
    </div>
  </div>
</DocumentFragment>
`;

exports[`TabsItem > should match a snapshot and render children 1`] = `
<DocumentFragment>
  <span
    class="foo el-tabs-item"
  >
    <div>
      Child
    </div>
  </span>
</DocumentFragment>
`;

exports[`TabsLabel > should match a snapshot and render children 1`] = `
<DocumentFragment>
  <label
    class="foo el-tabs-label"
  >
    <div>
      Child
    </div>
  </label>
</DocumentFragment>
`;

exports[`TabsOptionsWrap > should match a snapshot and render children 1`] = `
<DocumentFragment>
  <div
    class="foo el-tabs-options-wrap"
  >
    <div>
      Child
    </div>
  </div>
</DocumentFragment>
`;

exports[`TabsWrap > should match a snapshot and render children 1`] = `
<DocumentFragment>
  <div
    class="foo el-tabs-wrap"
  >
    <div>
      Child
    </div>
  </div>
</DocumentFragment>
`;
